File: SwitchController.swift
import AppKit
import Combine

/// SSID 변경을 구독하고, 매칭 프로필의 월페이퍼를 적용한다.
@MainActor
class SwitchController: ObservableObject {
    @Published private(set) var activeProfile: Profile?

    private let profileManager: ProfileManager
    private let locationWatcher: LocationWatcher
    private var cancellable: AnyCancellable?

    init(profileManager: ProfileManager, locationWatcher: LocationWatcher) {
        self.profileManager = profileManager
        self.locationWatcher = locationWatcher

        cancellable = locationWatcher.$currentGatewayMAC
            .receive(on: DispatchQueue.main)
            .sink { [weak self] mac in
                Task { @MainActor [weak self] in self?.onNetworkChange(mac) }
            }
    }

    /// 메뉴바 등에서 수동으로 프로필을 전환할 때 사용한다.
    func apply(_ profile: Profile) {
        applyProfile(profile)
    }

    private func onNetworkChange(_ mac: String?) {
        print("[SwitchController] 네트워크 변경 감지: \(mac ?? "nil")")
        guard let profile = profileManager.profileFor(gatewayMAC: mac) else {
            print("[SwitchController] ❌ 매칭 프로필 없음 (profiles: \(profileManager.profiles.map { "\($0.name)=\($0.gatewayMAC ?? "nil")" }))")
            return
        }
        print("[SwitchController] 매칭 프로필: \(profile.name) (wallpaper: \(profile.wallpaperPath.isEmpty ? "없음" : profile.wallpaperPath))")
        guard profile.id != activeProfile?.id else {
            print("[SwitchController] 이미 활성 프로필, 건너뜀")
            return
        }
        applyProfile(profile)
    }

    private func applyProfile(_ profile: Profile) {
        print("[SwitchController] 프로필 적용 시작: \(profile.name)")
        activeProfile = profile
        profileManager.activeProfileID = profile.id
        let path = profile.wallpaperPath
        guard !path.isEmpty else {
            print("[SwitchController] ⚠️ 월페이퍼 경로가 비어있음 — 프로필은 전환됐지만 월페이퍼 변경 없음")
            return
        }
        let ext = (path as NSString).pathExtension.lowercased()
        switch ext {
        case "mp4", "mov":
            applyVideo(path)
        case "jpg", "jpeg", "png":
            applyImage(path)
        default:
            break
        }
    }

    /// mp4/mov: WallpaperEngine으로 모든 디스플레이에 루프 재생
    private func applyVideo(_ path: String) {
        guard let engine = sharedEngine else { return }
        let displayIDs = NSScreen.screens.compactMap {
            $0.deviceDescription[NSDeviceDescriptionKey("NSScreenNumber")] as? NSNumber
        }
        engine.startWallpaper(withPath: path, onDisplays: displayIDs)
    }

    /// jpg/png: NSWorkspace로 정적 배경 설정, 기존 라이브 데몬 종료
    private func applyImage(_ path: String) {
        let url = URL(fileURLWithPath: path)
        for screen in NSScreen.screens {
            try? NSWorkspace.shared.setDesktopImageURL(url, for: screen, options: [:])
        }
        sharedEngine?.killAllDaemons()
    }
}

File: windows/Services/SwitchController.cs
using System.Diagnostics;
using Ilko.Models;

namespace Ilko.Services;

public class SwitchController
{
    private readonly ProfileManager _profileManager;
    private readonly LocationWatcher _locationWatcher;
    private readonly WallpaperEngine _engine;

    public Profile? ActiveProfile { get; private set; }
    public event Action<Profile?>? ActiveProfileChanged;

    public SwitchController(ProfileManager profileManager, LocationWatcher locationWatcher, WallpaperEngine engine)
    {
        _profileManager = profileManager;
        _locationWatcher = locationWatcher;
        _engine = engine;
        _locationWatcher.GatewayMACChanged += OnNetworkChange;
    }

    public void Apply(Profile profile)
    {
        if (profile.Id == ActiveProfile?.Id) return;
        ApplyProfile(profile);
    }

    public void ForceApply(Profile profile) => ApplyProfile(profile);

    public void ApplyCurrentNetwork()
    {
        var profile = _profileManager.ProfileFor(_locationWatcher.CurrentGatewayMAC);
        if (profile == null) return;
        ApplyProfile(profile);
    }

    private void OnNetworkChange(string? mac)
    {
        var profile = _profileManager.ProfileFor(mac);
        if (profile == null || profile.Id == ActiveProfile?.Id) return;
        ApplyProfile(profile);
    }

    private void ApplyProfile(Profile profile)
    {
        Debug.WriteLine($"[SwitchController] 적용: {profile.Name}");
        ActiveProfile = profile;
        _profileManager.ActiveProfileId = profile.Id;
        ActiveProfileChanged?.Invoke(profile);
        _engine.ApplyProfile(profile);
    }
}
